bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -282,97 +301,115 @@ object BagsManager {
-        case BagsState.RunningEditedState(_) => s"RunningEdited"
-        case BagsState.FailureState(_, why)  => s"Failure($why)"
-      }
-
-      val obsId = Option(obs.getObservationID).getOrElse(obs.getNodeKey)
-
-      Log.info(f"BAGS transition for obs $obsId: ${stateString(from)}%-30s - $label%-8s -> ${stateString(to)}")
-    }
-
-    val progKey = ProgKey(obs)
-    val obsKey  = ObsKey(obs)
-
-    val stateTuple = for {
-      obsMap <- stateMap.lookup(progKey)
+  private def transition(obsKey: ObsKey, update: BagsState => StateTransition): Unit = Swing.onEDT {
+    for {
+      obsMap <- stateMap.lookup(obsKey.prog)
       state  <- obsMap.lookup(obsKey)
-    } yield (obsMap, state)
-
-    stateTuple.foreach { case (obsMap, state) =>
+    } {
       val (newState, sideEffect) = update(state)
-      val newStatemap            = stateMap.insert(progKey, obsMap.insert(obsKey, newState))
+      val newStatemap            = stateMap.insert(obsKey.prog, obsMap.insert(obsKey, newState))
 
       // Some actions will produce further state transitions when executed and
       // those transitions assume the newState has been reached.  Therefore we
       // make sure to update the state map before running the associated actions.
       val action = for {
         _ <- IO { stateMap = newStatemap }
-        _ <- logTransition(state, newState)
         _ <- sideEffect
+        _ <- IO { fireBagsStateChanged(obsKey.key, state, newState) }
       } yield ()
-      action.unsafePerformIO()
 
-      fireBagsStateChanged(obs.getNodeKey, state, newState)
+      action.unsafePerformIO()
     }
   }
 
-  private def edited(obs: ISPObservation): Unit =
-    transition(obs, "edit", _.edit)
+  private def edited(o: ObsKey): Unit =
+    transition(o, _.edit)
 
-  private def wakeUp(obs: ISPObservation): Unit =
-    transition(obs, "wakeUp", _.wakeUp)
+  private def wakeUp(o: ObsKey): Unit =
+    transition(o, _.wakeUp)
 
-  private def success(obs: ISPObservation, results: Option[AgsStrategy.Selection]): Unit =
-    transition(obs, "succeed", _.succeed(results))
+  private def success(o: ObsKey, results: Option[AgsStrategy.Selection]): Unit =
+    transition(o, _.succeed(results))
 
-  private def fail(obs: ISPObservation, why: String, delay: Long): Unit =
-    transition(obs, "fail", _.fail(why, delay))
+  private def fail(o: ObsKey, why: String, delay: Long): Unit =
+    transition(o, _.fail(why, delay))
 
-  /** Add a program to our watch list and attach listeners. Enqueue all
+  /** Add a program to our watch list and attach listeners. Enqueue modified
     * observations for consideration for a BAGS lookup.
     */
   def watch(prog: ISPProgram): Unit = Swing.onEDT {
-    // Check that the program isn't being managed already. If it is, do nothing.
+    Log.info("watch")
+
+    // Remove existing listeners, if any.
+    prog.removeStructureChangeListener(StructureListener)
+    prog.removeCompositeChangeListener(ChangeListener)
```

Comment:
Is it possible to miss updates since we're not listening during the computation below?